### PR TITLE
Fix mul cuda for bool

### LINKED ISSUE #47007: RuntimeError: "mul_cuda" not implemented for 'Bool'
## 🐛 Bug

I just upgraded to PyTorch 1.7 and I now get an this error when multiplying a Python bool with a cuda bool tensor. I do not get the error on CPU. Furthermore, it was working well on the previous version (1.6).

## To Reproduce

image

## Environment

Collecting environment information...
PyTorch version: 1.7.0+cu110
Is debug build: True
CUDA used to build PyTorch: 11.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.1 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-10ubuntu2) 9.3.0
Clang version: Could not collect
CMake version: Could not collect

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: GeForce GTX 1080 Ti
GPU 1: GeForce GTX 1080 Ti
GPU 2: GeForce GT 1030

Nvidia driver version: 450.51.06
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.18.5
[pip3] torch==1.7.0+cu110
[pip3] torchaudio==0.7.0
[pip3] torchvision==0.8.1+cu110
[conda] Could not collect

### PR DESCRIPTION
**Pull request by malfet**
Also, add tests for tensor by scalar multiplication / division

Fixes https://github.com/pytorch/pytorch/issues/47007

### CODE REVIEW

File: aten/src/ATen/native/cpu/DistributionTemplates.h [hunk 1]
```diff
@@ -180,9 +180,7 @@ void normal_kernel(Tensor& self, double mean, double std, RNG generator) {
     normal_fill(self, static_cast<float>(mean), static_cast<float>(std), generator);
 #endif
   } else {
-    // bfloat16 cannot be properly tested due to the lack of other operations
-    // like add/sub/mean implemented for half
-    AT_DISPATCH_FLOATING_TYPES_AND_HALF(self.scalar_type(), "normal_kernel_cpu", [&] {
+    AT_DISPATCH_FLOATING_TYPES_AND2(kHalf, kBFloat16, self.scalar_type(), "normal_kernel_cpu", [&] {
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cpu/DistributionTemplates.h:183]: Thanks for fixing this. I was trying to create a bfloat tensor on some of my PR and this block me from doing so.

**Review by zasdfgbnm — COMMENTED** [review]
Generally LGTM, but I think you should also add `bfloat16` to `test_randn` in https://github.com/pytorch/pytorch/blob/master/test/test_tensor_creation_ops.py#L1204

File: aten/src/ATen/native/cuda/BinaryMulDivKernel.cu [hunk 2]
```diff
@@ -45,7 +45,7 @@ struct MulFunctor<bool> {
 
 
 void div_kernel_cuda(TensorIterator& iter) {
-  if (!isIntegralType(iter.common_dtype(), /*includeBool*/ false) && iter.is_cpu_scalar(2)) {
+  if (!isIntegralType(iter.common_dtype(), /*includeBool*/ true) && iter.is_cpu_scalar(2)) {
```
Comment (gchanan) [inline aten/src/ATen/native/cuda/BinaryMulDivKernel.cu:48]: does this affect anything?  neither of the dispatch macros has bool?

File: aten/src/ATen/native/cpu/DistributionTemplates.h (hunk 1 above)
Comment (gchanan) [reply aten/src/ATen/native/cpu/DistributionTemplates.h:183]: is this related to this change?  Test?

File: test/test_torch.py [hunk 3]
```diff
@@ -19359,6 +19359,22 @@ def compare_helper_(like_fn, t):
                     tp = t.permute(p)
                     compare_helper_(like_fn, tp)
 
+    @unittest.skipIf(TEST_WITH_ASAN, "Integer overflows are not allowed under ASAN")
+    @dtypes(*torch.testing.get_all_dtypes())
+    def test_muldiv_scalar(self, device, dtype):
+        if dtype.is_complex or dtype.is_floating_point:
+            x = torch.randn(10, 3, dtype=dtype, device=device)
```
Comment (gchanan) [inline test/test_torch.py:19366]: can you use `make_tensor` for this?

File: aten/src/ATen/native/cuda/BinaryMulDivKernel.cu (hunk 2 above)
Comment (malfet) [reply aten/src/ATen/native/cuda/BinaryMulDivKernel.cu:48]: You are right, it doesn't, which makes me a bit confused on what should be the expected behaviour here:
```
$ python -c "import torch;print(torch.tensor([True, False]) / True)"
tensor([1., 0.])
```
Although numpy behaves the same:
```
$ python -c "import numpy as np;print(np.asarray([True, False]) / True)"
[1. 0.]
```

File: aten/src/ATen/native/cpu/DistributionTemplates.h (hunk 1 above)
Comment (gchanan) [reply aten/src/ATen/native/cpu/DistributionTemplates.h:183]: make_tensor already has a solution for this right?  Using that and splitting up the PRs seems better.

File: test/test_torch.py [hunk 4]
```diff
@@ -19359,6 +19359,22 @@ def compare_helper_(like_fn, t):
...
+            s = torch.randn(1, dtype=dtype, device='cpu').item()
+        else:
+            dtype_max = torch.iinfo(dtype).max if dtype != torch.bool else 2
+            dtype_min = torch.iinfo(dtype).min if dtype != torch.bool else 0
+            x = torch.randint(dtype_min, dtype_max, (10, 3), device=device, dtype=dtype)
+            s = torch.randint(dtype_min, dtype_max, (1,), device='cpu', dtype=dtype).item()
+        y = torch.full_like(x, s)
+        self.assertEqual(x * s, x * y)
+        self.assertEqual(s * x, y * x)
+        self.assertEqual(x / s, x / y)
```
[the remark this answers is not in the document]
Comment (walterddr) [reply test/test_torch.py:19376]: divide operation is not commutative IMO.

**Review by walterddr — APPROVED** [review]

File: test/test_torch.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (malfet) [reply test/test_torch.py:19370]: Other order works, but only division of tensor by scalar is special cased (i.e. transformed into T * (1./s) ), but added both cases for completeness.

File: aten/src/ATen/native/cpu/DistributionTemplates.h (hunk 1 above)
Comment (malfet) [reply aten/src/ATen/native/cpu/DistributionTemplates.h]: Moved this change  to https://github.com/pytorch/pytorch/pull/47143

**Pull request closed without merging**